File: II-152-3/ProblemWriting.cpp
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE
template<typename T> string ToString( T x )
{
	ostringstream S; S << x; return S.str();
}
struct ProblemWriting
{
	enum S { s0, s1, s2, s3, s4 };
	string myCheckData( string dotForm )
	{
		if ( dotForm.empty() || dotForm.size() > 25 )
			return "dotForm must contain between 1 and 25 characters, inclusive.";
		int i = 0; S s = s0; string operators = "/*-+";
		for( ; i < dotForm.size(); ++i )
		{
			switch( s )
			{
				case s0:
					if ( ::isdigit( dotForm[i] ) ) s = s1;
					else s = s4;
					break;
				case s1:
					if ( operators.find( dotForm[i] ) != string::npos ) s = s3;
					else if ( dotForm[i] == '.' ) s = s2;
					else s = s4;
					break;
				case s2:
					if ( dotForm[i] == '.' ) s = s2;
					else if ( operators.find( dotForm[i] ) != string::npos ) s = s3;
					else s = s4;
					break;
				case s3:
					if ( dotForm[i] == '.' ) s = s3;
					else if ( ::isdigit( dotForm[i] ) ) s = s1;
					else s = s4;
					break;
			}
			if ( s == s4 ) break;
		}
		if ( s != s1 )
			return "dotForm is not in dot notation, check character " + ToString( i ) + ".";
		return "";
	}
	// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const long long& expected, const long long& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<string> A0; vector<string> A1;
	{ /*Test 0*/string dotForm = 
	"3+5"
	; A0.push_back( dotForm ); string result = 
	""
	; A1.push_back( result );  }
	{ /*Test 1*/string dotForm = 
	"9..+.5...*....3"
	; A0.push_back( dotForm ); string result = 
	""
	; A1.push_back( result );  }
	{ /*Test 2*/string dotForm = 
	"5.3+4"
	; A0.push_back( dotForm ); string result = 
	"dotForm is not in dot notation, check character 2."
	; A1.push_back( result );  }
	{ /*Test 3*/string dotForm = 
	"9*9*9*9*9*9*9*9*9*9*9*9*9*9"
	; A0.push_back( dotForm ); string result = 
	"dotForm must contain between 1 and 25 characters, inclusive."
	; A1.push_back( result );  }
	{ /*Test 4*/string dotForm = 
	"3.........../...........4"
	; A0.push_back( dotForm ); string result = 
	""
	; A1.push_back( result );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A1[i], myCheckData( A0[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	ProblemWriting ___test; 
	___test.run_test(); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE

File: II-152-3/ProblemWriting.cs
using System;
using System.Collections;
using System.Text;
using System.Text.RegularExpressions;

enum States { s0, s1, s2, s3 };

public class ProblemWriting
{
    public string myCheckData( string dotForm )
    {
		if ( dotForm.Length == 0 || dotForm.Length > 25 )
		{
			return "dotForm must contain between 1 and 25 characters, inclusive.";
		}
		States s = States.s0;
		int i = 0;
		while( i < dotForm.Length )
		{
			switch( s )
			{
				case States.s0:
					if ( dotForm[i] >= '0' && dotForm[i] <= '9' )
					{
						s = States.s1;
					}
					else
					{
						return "dotForm is not in dot notation, check character " + i.ToString() + ".";
					}
					break;
				case States.s1:
					if ( dotForm[i] == '.' )
					{
						s = States.s2;
					}
					else if ( dotForm[i] == '+' || dotForm[i] == '/' || dotForm[i] == '*' || dotForm[i] == '-' )
					{
						s = States.s3;
					}
					else
					{
						return "dotForm is not in dot notation, check character " + i.ToString() + ".";
					}
					break;
				case States.s2:
					if ( dotForm[i] == '.' )
					{
						s = States.s2;
					}
					else if ( dotForm[i] == '+' || dotForm[i] == '/' || dotForm[i] == '*' || dotForm[i] == '-' )
					{
						s = States.s3;
					}
					else
					{
						return "dotForm is not in dot notation, check character " + i.ToString() + ".";
					}
					break;
				case States.s3:
					if ( dotForm[i] >= '0' && dotForm[i] <= '9' )
					{
						s = States.s1;
					}
					else if ( dotForm[i] == '.' )
					{
						s = States.s3;
					}
					else
					{
						return "dotForm is not in dot notation, check character " + i.ToString() + ".";
					}
					break;
			}
			i++;
		}
		if ( s != States.s1 )
		{
			return "dotForm is not in dot notation, check character " + i.ToString() + ".";
		}
        return "";
    }

// BEGIN CUT HERE
  public static void Main(string[] args)
  {
    try
    {
		eq(1,(new ProblemWriting()).myCheckData("3+5"),"");
		eq(2,(new ProblemWriting()).myCheckData("9..+.5...*....3"),"");
		eq(3,(new ProblemWriting()).myCheckData("5.3+4"),"dotForm is not in dot notation, check character 2.");
		eq(4,(new ProblemWriting()).myCheckData("9*9*9*9*9*9*9*9*9*9*9*9*9*9"),"dotForm must contain between 1 and 25 characters, inclusive.");
		eq(5,(new ProblemWriting()).myCheckData("3.........../...........4"),"");
    }
    catch(Exception e)
    {
	Console.WriteLine(e.Message);
    }
  }
  private static void eq(int n, int a, int b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, char a, char b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
  }
  private static void eq(int n, long a, long b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, bool a, bool b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, string a, string b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, int[] a, int[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void eq(int n, long[] a, long[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void eq(int n, string[] a, string[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void print(int a)
  {
    Console.Write(a+" ");
  }
  private static void print(string s)
  {
    Console.Write("\""+s+"\" ");
  }
  private static void print(int[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write(rs[i]);
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void print(long[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write(rs[i]);
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void print(string[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write("\""+rs[i]+"\"");
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void nl()
  {
    Console.WriteLine();
  }
// END CUT HERE
}
